botocore/handlers.py

```diff
@@ -140,6 +143,28 @@ def conditionally_calculate_md5(params, **kwargs):
         calculate_md5(params, **kwargs)
 
 
+def validate_bucket_name(params, **kwargs):
+    # The rules for bucket names in the US Standard region allow bucket names
+    # to be as long as 255 characters, and bucket names can contain any
+    # combination of uppercase letters, lowercase letters, numbers, periods
+    # (.), hyphens (-), and underscores (_).
+    if 'Bucket' not in params:
+        return
+    bucket = params['Bucket']
+    if len(bucket) > 255:
+        error_msg = (
+            'Invalid bucket name "%s": Bucket name cannot be '
+            'longer than 255 characters.' % bucket)
+        raise ParamValidationError(report=error_msg)
+    for char in bucket:
```

Comment:
The main reason for not using the regex was just to have more specific error messages for the two cases.  Otherwise we'd just have to say it's not valid, and maybe just show the regex it has to match.